core/embed/rust/src/ui/display/toif.rs

```diff
@@ -0,0 +1,183 @@
+    if let Ok(info) = trezorhal::display::toif_info(data) {
+        Some((
+            Offset::new(
+                unwrap!(info.width.try_into()),
+                unwrap!(info.height.try_into()),
+            ),
+            info.format,
+        ))
+    } else {
+        None
+    }
+}
+
+pub fn icon_precise(icon: &Icon, center: Point, fg_color: Color, bg_color: Color) {
+    let r = Rect::from_center_and_size(center, icon.size());
+
+    let area = r.translate(get_offset());
+    let clamped = area.clamp(constant::screen());
+    let colortable = get_color_table(fg_color, bg_color);
+
+    set_window(clamped);
+
+    let mut dest = [0_u8; 1];
+
+    let mut window = [0; UZLIB_WINDOW_SIZE];
+    let mut ctx = UzlibContext::new(icon.data(), Some(&mut window));
+
+    for py in area.y0..area.y1 {
+        for px in area.x0..area.x1 {
+            let p = Point::new(px, py);
+            let x = p.x - area.x0;
+
+            if clamped.contains(p) {
+                if x % 2 == 0 {
+                    unwrap!(ctx.uncompress(&mut dest), "Decompression failed");
+                    pixeldata(colortable[(dest[0] & 0xF) as usize]);
+                } else {
+                    pixeldata(colortable[(dest[0] >> 4) as usize]);
+                }
+            } else if x % 2 == 0 {
+                //continue unzipping but dont write to display
+                unwrap!(ctx.uncompress(&mut dest), "Decompression failed");
+            }
+        }
+    }
+
+    pixeldata_dirty();
+}
+
+/// Holding toif data and allowing it to draw itself.
+#[derive(PartialEq, Eq, Clone, Copy)]
+pub struct Toif {
+    data: &'static [u8],
+    name: &'static str, // useful for debugging purposes.
+    size: Offset,
+    format: ToifFormat,
+}
+
+impl Toif {
+    pub fn new(named_toif: NamedToif) -> Self {
+        let info = unwrap!(toif_info(named_toif.0));
+        Self {
+            data: named_toif.0[TOIF_HEADER_LENGTH..].as_ref(),
+            name: named_toif.1,
+            size: info.0,
+            format: info.1,
+        }
+    }
+
+    pub fn from_slice(data: &'static [u8]) -> Self {
+        let named_toif = NamedToif(data, "");
+        Self::new(named_toif)
+    }
+
+    pub fn ensure(&self, format: ToifFormat) {
+        assert_eq!(self.format, format);
+    }
+
+    pub fn format(&self) -> ToifFormat {
+        self.format
+    }
+
+    pub fn name(&self) -> &'static str {
+        self.name
+    }
+}
+
+pub trait ToifAccessors {
+    fn toif(&self) -> Toif;
+
+    fn data(&self) -> &[u8] {
+        self.toif().data
+    }
+
+    fn format(&self) -> ToifFormat {
+        self.toif().format
+    }
+
+    fn width(&self) -> i16 {
+        self.toif().size.x
+    }
+
+    fn height(&self) -> i16 {
+        self.toif().size.y
+    }
+
+    fn size(&self) -> Offset {
+        self.toif().size
+    }
+}
+
+#[derive(PartialEq, Eq, Clone, Copy)]
+pub struct Icon {
+    toif: Toif,
+    area: Rect,
```

Comment:
area is unused for `Icon` because it's not a component

Reply:
removed https://github.com/trezor/trezor-firmware/commit/f05d93194cb77e02f35a87353a3614cd85fb1dbe